packages/block-library/src/image/index.php

```diff
@@ -19,10 +19,12 @@ function render_block_core_image( $attributes, $content ) {
 		// to provide backwards compatibility for the Gallery Block,
 		// which now wraps Image Blocks within innerBlocks.
 		// The data-id attribute is added in a core/gallery `render_block_data` hook.
-		$data_id_attribute = 'data-id="' . esc_attr( $attributes['data-id'] ) . '"';
-		if ( false === strpos( $content, $data_id_attribute ) ) {
-			$content = str_replace( '<img', '<img ' . $data_id_attribute . ' ', $content );
+		$walker = new WP_HTML_Walker( $content );
+		$walker->next_tag( 'img' );
+		if ( $walker->get_attribute( 'data-id' ) !== $attributes['data-id'] ) {
+			$walker->set_tag_attribute( 'data-id', $attributes['data-id'] );
```

Comment:
`$walker->set_attribute` no?

Reply:
this is a good place to examine the escaping. I think after reading some parsing material the other day we should be able to safely _always_ escape all attribute values, even CSS class names. right here this would keep the interface we expect, but we'd have to make sure that in `get_attribute()` we perform the reverse-escaping.

could cause trouble when comparing equality here since we don't know if sanitization will change things.

Reply:
> `$walker->set_attribute` no?
I didn't run the code yet as I wanted to share the changes so we could see how API plays in practice 😅

that's a great point about the comparisons between escaped and unescaped values, we could also leave it up to the developers 

```js
if ( $walker->get_attribute( 'data-id' ) !== esc_attr( $attributes['data-id'] ) ) {
```

Reply:
I'd rather not leave it up to every developer because I feel like that's how things get inconsistent and unreliable. If the interface is a PHP string value we should be able to make the right conditions easy; the default approach (of not doing anything) correct.